fairscale/nn/misc/grad_bucket.py

```diff
@@ -0,0 +1,148 @@
+
+        self.params_checked_in = 0
+        self.destination = destination
+        self.process_group = process_group
+        self.sent = True
+        self.callback: Optional[Callable[[Any], None]] = None
+
+    def reset_checked_in(self) -> None:
+        """ Reset the counter of the parameter grads which have been checked in
+        """
+        self.params_checked_in = 0
+        self.sent = False
+
+    @property
+    def all_checked_in(self) -> bool:
+        """ Have all the expected gradient check-in happened ?"""
+        return len(self._params) == self.params_checked_in
+
+    def can_add_grad_view(self, param: torch.Tensor) -> bool:
+        """ Is there enough room in the bucket to add this parameter gradient ?
+        """
+        return self._fill + param.numel() < self._max_size
+
+    def to(  # type: ignore
+        self,
+        device: Optional[Union[int, torch.device]],
+        dtype: Optional[torch.dtype] = None,
+        non_blocking: bool = False,
+    ) -> "GradBucket":
+        """
+        Move the underlying buffer
+        """
+        if self._is_collapsed:
+            self.rebuild()
+
+        assert self._buffer is not None, "Cannot move a collapsed bucket, please rebuild it"
+        self._buffer.to(device, dtype, non_blocking)
+
+    def zero(self) -> None:
+        """
+        Set all the grads to zero
+        """
+        if self._buffer is not None:
+            self._buffer.fill_(0.0)
+
+    @torch.no_grad()
+    def add_grad(self, param: torch.Tensor) -> None:
+        """
+        Add a new parameter gradient to the bucket. Param.grad becomes a view of this bucket buffer
+        """
+        if param.grad is None:
+            param.grad = torch.zeros_like(param)
+
+        self._add_grad_as_view(param)
+        self._params.append(param)
+
+    @torch.no_grad()
+    def collapse(self) -> None:
+        """
+        Release the buffer from memory. The bucket will need to be rebuilt before use
+        """
+        if not self._is_collapsed:
+            for p in self._params:
+                p.grad = None
+
+            self._buffer = None
+            self._fill = 0
+            self.params_checked_in = 0
+            self._is_collapsed = True
+
+    @torch.no_grad()
+    def rebuild(self) -> None:
+        """
+        Given the parameter gradients which have been registered previously, rebuild the whole bucket
+        """
+        assert len(self._params) > 0
+
+        if self._is_collapsed:
+            self._buffer = torch.zeros(self._max_size, dtype=self._params[0].dtype, device=self._params[0].device)
+
+            for p in self._params:
+                self._add_grad_as_view(p)
+
+            self._is_collapsed = False
+
+    @torch.no_grad()
+    def scale_and_reduce(self, scale: float) -> torch.futures.Future:
+        assert self._buffer is not None, "Cannot send a collapsed bucket, please rebuild"
+
+        self._buffer.mul_(scale)
+        torch.cuda.synchronize()
```

Comment:
this is very expensive in terms of time, but the results are not correct without it. If a reviewer can think of a better synchronization that would be great..